lib/sycamore/sycamore/utils/time_trace.py

```diff
@@ -4,8 +4,59 @@
 import resource
 import threading
 import functools
+import logging
 from sys import platform
 
+zero_time = time.time_ns()
+# Mac reports ru_maxrss in bytes. Linux reports it in kB.
+if platform == "darwin":
+    RSS_MULTIPLIER = 1
+else:
+    RSS_MULTIPLIER = 1024
+
+
+class TimeTraceData:
+    def __init__(self, t0, t1, user, sys, rss):
+        self.t0 = t0  # ns
+        self.t1 = t1  # ns
+        self.user = user  # ns
+        self.sys = sys  # ns
+        self.rss = rss  # bytes
+
+    def wall_s(self):
+        return (self.t1 - self.t0) / 1.0e9
+
+    def user_s(self):
+        return self.user / 1.0e9
+
+    def sys_s(self):
+        return self.sys / 1.0e9
+
+    def rss_mib(self):
+        return self.rss / (1024 * 1024)
+
+
+class InMemoryTimeTrace:
+    try:
+        resource_type = resource.RUSAGE_THREAD  # type: ignore
+    except AttributeError:
+        resource_type = resource.RUSAGE_SELF
+
+    def __init__(self):
+        self.t0 = time.time_ns()
+        self.r0 = resource.getrusage(self.resource_type)
+
+    def end(self):
+        t1 = time.time_ns()
+        r1 = resource.getrusage(self.resource_type)
+        r0 = self.r0
+        user = int((r1.ru_utime - r0.ru_utime) * 1.0e9)
+        sys = int((r1.ru_stime - r0.ru_stime) * 1.0e9)
+
+        rss = r1.ru_maxrss * RSS_MULTIPLIER  # could do max(r0,r1), but this is more specific
+
+        return TimeTraceData(self.t0, t1, user, sys, rss)
```

Comment:
It's a bit surprising to have end() return computed data.  Maybe a name more like "reap" would be clearer.  There's no start() that needs and end() counterpoint.

Reply:
changed to .measure()